Source/JSONAPISerialization.swift

```diff
@@ -0,0 +1,164 @@
+//
+//  JSONAPISerialization.swift
+//  KakapoExample
+//
+//  Created by Alex Manzella on 28/04/16.
+//  Copyright © 2016 devlucky. All rights reserved.
+//
+
+import Foundation
+
+/// A protocol to serialzie entities conforming to JSON API.
+public protocol JSONAPISerializable {
+    /**
+     Builds the `data` field conforming to JSON API
+     
+     - parameter includeRelationships: Defines if it should include the `relationships` field
+     - parameter includeAttributes:    Defines if it should include the `attributes` field
+     
+     - returns: Return an object representing the `data` field conforming to JSON API
+     */
+    func data(includeRelationships includeRelationships: Bool, includeAttributes: Bool) -> AnyObject?
+}
+
+/**
+ *  A JSON API entity, conforming to this protocol will change the behavior of serialization, `CustomSerializable` behavior will be overriden by the JSON API behavior.
+ *  Relationships of an entity should also conform to this protocol, relationships are automatically recognized using the staic type of the property. For example an Array of `JSONAPIEntity` would be recognized as a relationship, also when empty, as soon as is static type at compile time is inferred correctly.
+ 
+    ```swift
+        struct User: JSONAPIEntity {
+            let friends: [Friend] // correct if friend is JSONAPIEntity
+            let enemies: [Any] // incorrect, Any is not JSONAPIEntity so this property is an attribute instead of a relationship
+        }
+    ```
+ */
+public protocol JSONAPIEntity: CustomSerializable, JSONAPISerializable {
```

Comment:
Overall the explanations regarding JSONAPI are good, but I would also add some references to the official JSONAPI when possible (e.g for the `relationships` and `attributes` parameters, or how are the relationships included within data)